integration_test/ops_agent_test/main_test.go

```diff
@@ -5102,6 +5114,159 @@ func TestLogCompression(t *testing.T) {
 	})
 }
 
+// listAndDeleteResources lists all gcloud resources of a given resourceType that match the gcloudFilter.
+// All the found listed resources are then deleted sequentially.
+// extraArguments can be used for different type of gcloud command requirements.
+func listAndDeleteResources(ctx context.Context, logger *log.Logger, resourceType []string, gcloudFilter string, extraArguments []string) error {
+	type gcloud_resource struct {
+		Name string
+		Zone string
+	}
+
+	listArgs := append(resourceType,
+		[]string{"list",
+			"--filter=" + gcloudFilter,
+			"--format=json",
+		}...)
+	listArgs = append(listArgs, extraArguments...)
+
+	output, err := gce.RunGcloud(ctx, logger, "", listArgs)
+	if err != nil {
+		return err
+	}
+
+	var resources []gcloud_resource
+	if err := json.Unmarshal([]byte(output.Stdout), &resources); err != nil {
+		return fmt.Errorf("could not parse JSON from %q: %v", output.Stdout, err)
+	}
+
+	// Nothing to delete.
+	if len(resources) == 0 {
+		return nil
+	}
+
+	for _, r := range resources {
+		deleteArgs := append(resourceType,
+			[]string{"delete", filepath.Base(r.Name),
+				"--format=json",
+			}...)
+		if r.Zone != "" {
+			deleteArgs = append(deleteArgs, "--zone")
+			deleteArgs = append(deleteArgs, filepath.Base(r.Zone))
+		}
+
+		deleteArgs = append(deleteArgs, extraArguments...)
+		_, err = gce.RunGcloud(ctx, logger, "", deleteArgs)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+const AppHubIntegrationTestApp = "ops-agent-app-hub-integration-test-app"
+
+func cleanupStaleResourcesForTestAppHubLogLabels(ctx context.Context, logger *log.Logger) error {
+	project := os.Getenv("PROJECT")
+	// 24h old resources are considered stale
+	staleResourceTime := time.Now().Add(-24 * time.Hour)
+
+	// Formated timestamp
+	staleResourceTimestamp := staleResourceTime.Format(time.RFC3339)
+
+	listAndDeleteResources(ctx, logger, []string{"compute", "instance-groups", "managed"},
+		"( creationTimestamp < "+staleResourceTimestamp+" AND name ~ .*test-[0-9]{1,8}-.*-mig$ )",
+		[]string{"--project", project})
+	listAndDeleteResources(ctx, logger, []string{"compute", "instance-templates"},
+		"( creationTimestamp < "+staleResourceTimestamp+" AND name ~ .*test-[0-9]{1,8}-.*-temp$ )",
+		[]string{"--project", project})
+	listAndDeleteResources(ctx, logger, []string{"apphub", "applications", "workloads"},
+		"( createTime < "+staleResourceTimestamp+" AND name ~ .*test-[0-9]{1,8}-.*-wl$ )",
+		[]string{"--project", project, "--application", AppHubIntegrationTestApp, "--location", "global"})
+
+	return nil
+}
+
+func TestAppHubLogLabels(t *testing.T) {
+	t.Parallel()
+	gce.RunForEachImage(t, func(t *testing.T, imageSpec string) {
+		t.Parallel()
+
+		ctx, logger, migVM := setupMainLogAndManagedInstaceGroupVM(t, imageSpec)
+
+		// Cleanup stale resources.
+		t.Cleanup(func() { cleanupStaleResourcesForTestAppHubLogLabels(ctx, logger) })
```

Comment:
You probably don't want to do this for each image, because they'll all race with each other and try to delete the same resources multiple times. (I think it might be sufficient to just hoist this up to outside the `gce.RunForEachImage` call)

Reply:
You are right. I updated the code to only run `cleanupStaleResourcesForTestAppHubLogLabels` once per execution of the `TestAppHubLogLabels` test, while keeping all logs in the same folder. This by only running the cleanup in the first imageSpec given to the test.